tests/Integration/Adapter/ContextStateManagerTest.php

```diff
@@ -205,4 +217,50 @@ public function testShopStateGroup(): void
         $this->assertEquals(Shop::CONTEXT_GROUP, Shop::getContext());
         $this->assertNull($contextStateManager->getContextFieldsStack());
     }
+
+    public function testShopContextState(): void
+    {
+        $this->legacyContext->getContext()->shop = $this->basicShop;
+
+        Shop::setContext(Shop::CONTEXT_SHOP, $this->basicShop->id);
+        $this->assertEquals($this->basicShop->id, $this->legacyContext->getContext()->shop->id);
+        $this->assertEquals($this->basicShop->id, Shop::getContextShopID());
+        $this->assertEquals($this->basicShop->id_shop_group, Shop::getContextShopGroupID());
+        $this->assertEquals(Shop::CONTEXT_SHOP, Shop::getContext());
+
+        $contextStateManager = new ContextStateManager($this->legacyContext);
+        $this->assertNull($contextStateManager->getContextFieldsStack());
+
+        $contextStateManager->setShopContext(ShopConstraint::shop($this->shop2->id));
+        $this->assertEquals($this->shop2->id, $this->legacyContext->getContext()->shop->id);
+        $this->assertEquals($this->shop2->id, Shop::getContextShopID());
+        $this->assertEquals($this->shop2->id_shop_group, Shop::getContextShopGroupID());
+        $this->assertEquals(Shop::CONTEXT_SHOP, Shop::getContext());
+        $this->assertCount(1, $contextStateManager->getContextFieldsStack());
+
+        $contextStateManager->setShopContext(ShopConstraint::shopGroup($this->shop3->id_shop_group));
+        // current shop itself doesn't change, (just the context) so we expect previous shop id here
+        $this->assertEquals($this->shop2->id, $this->legacyContext->getContext()->shop->id);
+        $this->assertNull(Shop::getContextShopID());
+        $this->assertEquals($this->shop3->id_shop_group, Shop::getContextShopGroupID());
+        $this->assertEquals(Shop::CONTEXT_GROUP, Shop::getContext());
+        $this->assertCount(1, $contextStateManager->getContextFieldsStack());
+
+        $contextStateManager->setShopContext(ShopConstraint::allShops());
+        // current shop itself doesn't change (just the context), so we expect previous shop id here
+        $this->assertEquals($this->shop2->id, $this->legacyContext->getContext()->shop->id);
+        $this->assertNull(Shop::getContextShopID());
+        $this->assertNull(Shop::getContextShopGroupID());
+        $this->assertEquals(Shop::CONTEXT_ALL, Shop::getContext());
+        $this->assertCount(1, $contextStateManager->getContextFieldsStack());
+
+        //@todo: I don't understand why this passes.
+        //      Shouldn't it restore context one step back? to the "allShops" and not all the way back to basic shop?
+        //      Also why there are no calls of "saveCurrentContext" it seemed to be important, I don't know what to expect.
```

Comment:
ping @Progi1984 @jolelievre 
added this integration test but I don't really know what to expect, it doesn't seem right this way, could you help to understand the expected behavior? (see todo above)

Reply:
ok then i remove todo and leave it.